usr.bin/ldd/ldd.c

```diff
@@ -385,106 +455,100 @@ is_executable(const char *fname, int fd, int *is_shlib, int *type,
-				break;
-			}
+		elf = elf_begin(fd, ELF_C_READ, NULL);
+		if (elf == NULL) {
+			warnx("%s: %s", fname, elf_errmsg(0));
+			return (0);
 		}
-
-		if (!dynamic) {
+		if (elf_kind(elf) != ELF_K_ELF) {
+			elf_end(elf);
 			warnx("%s: not a dynamic ELF executable", fname);
 			return (0);
 		}
-		if (hdr.elf32.e_type == ET_DYN) {
-			if (hdr.elf32.e_ident[EI_OSABI] == ELFOSABI_FREEBSD) {
-				*is_shlib = 1;
-				return (1);
-			}
-			warnx("%s: not a FreeBSD ELF shared object", fname);
+		if (gelf_getehdr(elf, &ehdr) == NULL) {
+			warnx("%s: %s", fname, elf_errmsg(0));
+			elf_end(elf);
 			return (0);
 		}
 
-		return (1);
-	}
-#endif
-
-	if ((size_t)n >= sizeof(hdr.elf) && IS_ELF(hdr.elf) &&
-	    hdr.elf.e_ident[EI_CLASS] == ELF_TARG_CLASS) {
-		/* Handle default ELF objects on this architecture */
-		Elf_Phdr phdr;
-		int dynamic, i;
-
-		dynamic = 0;
 		*type = TYPE_ELF;
 		*rtld = _DEFAULT_PATH_RTLD;
-
-		if (lseek(fd, hdr.elf.e_phoff, SEEK_SET) == -1) {
-			warnx("%s: header too short", fname);
-			return (0);
+#if __ELF_WORD_SIZE > 32 && defined(ELF32_SUPPORTED)
+		if (gelf_getclass(elf) == ELFCLASS32) {
+			*type = TYPE_ELF32;
+			*rtld = _COMPAT32_PATH_RTLD;
 		}
-		for (i = 0; i < hdr.elf.e_phnum; i++) {
-			if (read(fd, &phdr, hdr.elf.e_phentsize)
-			   != sizeof(phdr)) {
-				warnx("%s: can't read program header", fname);
+#endif
+
+		for (i = 0; i < ehdr.e_phnum; i++) {
+			if (gelf_getphdr(elf, i, &phdr) == NULL) {
+				warnx("%s: %s", fname, elf_errmsg(0));
+				elf_end(elf);
 				return (0);
 			}
-			if (phdr.p_type == PT_DYNAMIC) {
-				dynamic = 1;
+			switch (phdr.p_type) {
+			case PT_DYNAMIC:
+				dynamic = true;
+				break;
+			case PT_INTERP:
+				interp = true;
 				break;
 			}
 		}
 
 		if (!dynamic) {
+			elf_end(elf);
 			warnx("%s: not a dynamic ELF executable", fname);
 			return (0);
 		}
-		if (hdr.elf.e_type == ET_DYN) {
-			switch (hdr.elf.e_ident[EI_OSABI]) {
-			case ELFOSABI_FREEBSD:
-				*is_shlib = 1;
-#ifdef __mips
-				if ((hdr.elf.e_flags & EF_MIPS_ABI) ==
-				    EF_MIPS_ABI_CHERIABI)
-					*rtld = _CHERIABI_PATH_RTLD;
 
+		/*
+		 * PIE binaries are ET_DYN, so use the presence of
+		 * PT_INTERP to differentiate executables from shared
+		 * libraries.
+		 */
+		if (!interp)
+			*is_shlib = 1;
+
+		if (!is_freebsd_elf(fname, elf, &ehdr)) {
+#if defined(__aarch64__) || defined(__riscv)
+			/*
+			 * Shared libraries on AArch64 and RISC-V have
+			 * neither a FreeBSD OSABI or a brand note.
+			 */
+			if (interp)
 #endif
-				return (1);
-#ifdef __ARM_EABI__
-			case ELFOSABI_NONE:
-				if (hdr.elf.e_machine != EM_ARM)
-					break;
-				if (EF_ARM_EABI_VERSION(hdr.elf.e_flags) <
-				    EF_ARM_EABI_FREEBSD_MIN)
-					break;
-				*is_shlib = 1;
-				return (1);
-#endif
+			{
+				elf_end(elf);
+				warnx("%s: not a FreeBSD ELF file", fname);
+				return (0);
 			}
-			warnx("%s: not a FreeBSD ELF shared object", fname);
-			return (0);
 		}
+		elf_end(elf);
 
+#ifndef __CHERI_PURE_CAPABILITY__
```

Comment:
Shouldn't this be checking if compat cheriabi is set? I guess checking the ABI of ldd works. However, I think we still force mips64 for cheribsdbox so it will break when embedded there.

Reply:
Probably it should be COMPAT_CHERIABI, yes.

Reply:
Actually, no the use case is not a lib64 ldd (ldd64 can mostly die I think since we don't have the execldd64 hack), this is needed for the hybrid world.